### PR TITLE
Review edits for PR 78

### PR DESCRIPTION
**Pull request by pllim**
My edits for your spacetelescope/acstools#78 . Edits include but not limited to:

* Consistent use of the logging system.
* Refactor some check logic.
* Documentation formatting.
* Future-proofing ease of maintenance.
* Miscellaneous nit-picking.

I let my OCD get the better of me sometimes. Thank you for your patience!

### CODE REVIEW

File: acstools/acszpt.py
Comment (pllim) [inline acstools/acszpt.py:76]: `__all__[0]` value may change in the future. You never know!

File: acstools/acszpt.py
Comment (pllim) [inline acstools/acszpt.py]: I have a feeling this might generate Sphinx warnings to have unbulleted sub-list. That's why.

File: acstools/acszpt.py
Comment (pllim) [inline acstools/acszpt.py:112]: I still think `**Attribute**` mention is unnecessary. If it is not rendering as you expect, there may be deeper issue in the doc build. I'll be happy to have a look if that is the case.

File: acstools/acszpt.py
Comment (pllim) [inline acstools/acszpt.py:152]: No point to recreate this string over and over again below.

File: acstools/acszpt.py
Comment (pllim) [inline acstools/acszpt.py:177]: The big refactoring happened here. I eliminated some loops and have the logging behavior to be simpler and more consistent. There is no need to have both logging and `print` to happen in the same logic block. Also not sure why you inserted leading whitespace in the messages. Better to leave that to the logger.

If you really want to, you can format the logger behavior up-front: https://docs.python.org/3/library/logging.html#formatter-objects

File: acstools/acszpt.py
Comment (pllim) [inline acstools/acszpt.py:235]: Printing the date here might better let user see why it does not match.

File: acstools/acszpt.py
Comment (pllim) [inline acstools/acszpt.py:237]: Logic block like this can be avoided most of the time, as in this case.
```python
if something:  # unnecessary
    pass
else:
    the_real_stuff
```

File: acstools/acszpt.py
Comment (pllim) [inline acstools/acszpt.py:252]: It can be confusing to mix error and info message in the same logic block. If the check failed, the entire message should go into error.

File: acstools/acszpt.py
Comment (pllim) [inline acstools/acszpt.py:284]: I think there might actually be two different magic numbers here, but just that they both happen to be 6? But I don't understand the in-line comment enough to be 100% sure.

File: acstools/acszpt.py
Comment (pllim) [inline acstools/acszpt.py:286]: Tsk tsk, double loop. If in the future, this turns out to be a bottleneck, you might need to revise a more efficient way to parse. I'll leave it be for now.

File: acstools/acszpt.py
Comment (pllim) [inline acstools/acszpt.py:328]: Don't really need to point out the obvious? 😄

File: acstools/acszpt.py
Comment (pllim) [inline acstools/acszpt.py:355]: Plain `return` is same as returning `None`.

File: acstools/satdet.py
Comment (pllim) [inline acstools/satdet.py:132]: This is some future-proofing that I am sneaking in here. 😉 Otherwise, it is more work as a new PR, which then we'll have to rebase on top of each other.

**Review by pllim — COMMENTED** [review]
I left some reasoning on why I did the things that I did. I have to admit that I did not test this out. So, please go ahead and give it a spin, or let me know of any questions or concerns. If you disagree on some of the changes, we can totally work it out; no problem.

p.s. Did you really test this in Python 2? The `urllib` import and `print` would have failed in Python 2 before this patch.

File: acstools/acszpt.py
Comment (nmiles2718) [reply acstools/acszpt.py]: It renders correctly when I build things locally.

File: acstools/acszpt.py
Comment (nmiles2718) [reply acstools/acszpt.py:112]: I'll swing by your office on Monday and I can show you what I mean.

File: acstools/acszpt.py
Comment (nmiles2718) [reply acstools/acszpt.py:177]: I always forget about the `join` method! Well done 👏 lol

File: acstools/acszpt.py
Comment (nmiles2718) [reply acstools/acszpt.py:284]: No, there is only one magic number. It corresponds to the number of columns in the `HTML` table. The outer loop shifts the starting index so that the first value in the inner loop is the column name. Then every 6th element corresponds to a member of that column.

File: acstools/acszpt.py
Comment (nmiles2718) [reply acstools/acszpt.py:286]: I figured out a better way to do it with `numpy` and no loops 😛 . I think I'll make a `dict` to store the units in and then we can just loop over the `dict` to assign units to each column without the need to have the 6 `if/elif` statements. Once that is done we can just combine the `_parse_response()` and `_format_results()` methods into a single method. What do you think?

```console
In [68]: q._submit_request()                                                                                            

In [69]: soup = BeautifulSoup(q._response.read(), 'html.parser')                                                        

In [70]: td = soup.find_all('td')                                                                                       

In [71]: td = [val.text.split(' ')[0] for val in td]                                                                    

In [72]: a = np.reshape(td, (int(len(td)/BLOCK_SIZE), BLOCK_SIZE))                                                      

In [73]: a[0]                                                                                                           
Out[73]: 
array(['FILTER', 'PHOTPLAM', 'PHOTFLAM', 'STmag', 'VEGAmag', 'ABmag'],
      dtype='<U9')

In [74]: a[1]                                                                                                           
Out[74]: 
array(['F220W', '2255.6', '8.166e-18', '21.620', '21.870', '23.546'],
      dtype='<U9')

In [75]: tb = Table(a[1, :], names = a[0], dtype = [str, float, float, float, float, float])

In [76]: tb                                                                                                             
Out[76]: 
<Table length=17>
FILTER PHOTPLAM  PHOTFLAM  STmag  VEGAmag  ABmag 
 str9  float64   float64  float64 float64 float64
------ -------- --------- ------- ------- -------
 F220W   2255.6 8.166e-18   21.62   21.87  23.546
 F250W   2716.1 4.673e-18  22.226  22.255  23.748
 F330W   3362.8  2.22e-18  23.034  22.908  24.093
 F344N   3433.8 2.209e-17   20.54    20.4  21.553
 F435W   4310.8 5.333e-19  24.583  25.185  25.102
 F475W   4775.2 2.893e-19  25.247  25.632  25.544
 F502N   5022.9 8.003e-18  21.642  21.909  21.829
 F550M   5579.7 5.818e-19  24.488  24.417  24.447
 F555W   5355.5 2.995e-19  25.209  25.258  25.257
 F606W   5886.5 1.262e-19  26.147  25.903   25.99
 F625W   6295.0 1.949e-19  25.675  25.206  25.372
 F658N   6583.7 3.333e-18  22.593   21.81  22.193
 F660N   6599.0 8.787e-18   21.54  20.849  21.135
 F775W   7664.7  1.93e-19  25.686  24.566  24.955
 F814W   8114.9  1.27e-19   26.14  24.854  25.286
F850LP   9144.0  2.25e-19   25.52  23.878  24.406
 F892N   8916.1 2.439e-18  22.932  21.379  21.873
```

**Comment by nmiles2718** [comment]
No, I didn't test it on python 2 because we shouldn't be encouraging our users to use python 2 by providing them with code that is compatible with python 2. I understand all of the modules we currently have in `acstools` are already compatible with 2 and 3, but I don't see the point in continuing that with newer modules in light of the fact that python 2.7 will be officially unsupported in ~11 months.

File: acstools/acszpt.py
Comment (pllim) [reply acstools/acszpt.py:286]: Nice! But is there any chance at this point that the query would return empty table (instead of failing in the checks)? If there is a chance, this would give `IndexError`.

File: acstools/acszpt.py
Comment (pllim) [reply acstools/acszpt.py:112]: Okay, I am curious!

File: acstools/acszpt.py
Comment (pllim) [reply acstools/acszpt.py:284]: Something to consider for the future -- What is the maintenance cost if the backend changes on your server? Look at `astroquery`, they have rolling release now because the backends are always changing on them. I don't think `acstools` will want to go this route. If there is a way to parse this without that magic number that is resistant to backend changes, it would be better.

**Comment by pllim** [comment]
Re: Python 2 -- I agree with you! Unfortunately, `acstools` as a package (along with some others) is lagging behind. If I did not change the import, it will mess up our Astroconda distribution in the Python 2 environment, and then I'll have to explain to everyone why I allowed it to break. So, just dealing with the compatibility for now is much less painful for me. We will definitely come back and remove all PY2 compat stuff to the whole package when Harry or Steve or whoever gives the go-ahead. The `tools` management chain is not very clear right now.

**Comment by pllim** [comment]
If you really, really don't want this to run in Python 2, then the alternative is adding a Python version check in the class constructor and raise an exception when someone tries to run this in Python 2. And move any incompatible imports into the method level. What say you?

**Comment by nmiles2718** [comment]
Actually, handling Python 2 and 3 might be easier than we think. We can leverage the fact that in either case the packages used are native ones and so they will _always_ be installed. The following works so long as the appropriate code changes are made in the `_submit_request()` method:
``` python 
try:
    from urllib.request import urlopen
    from error import URLError
except ImportError:
    from urllib2 import urlopen, URLError
```
I've got it working on my branch and I've tested it in python 2.7 and it works fine.

**Comment by pllim** [comment]
Re: https://github.com/nmiles2718/acstools/pull/1#issuecomment-457979053

This should not be necessary with the `six` imports.

File: acstools/acszpt.py
Comment (pllim) [reply acstools/acszpt.py:112]: As discussed, we agreed to move these to properties and rely on sphinx-automodapi to render them nicely.

**Review by nmiles2718 — APPROVED** [review]
Looks good!

File: acstools/acszpt.py
Comment (pllim) [inline acstools/acszpt.py:111]: This invoked the http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#definition-lists rendering. FYI.

**Review by pllim — COMMENTED** [review]
This should be ready for merge if you are okay with it. As discussed, after this is merged, you can proceed with:

* Changing usage of `Table` to `QTable`, so that extracted value is `Quantity` instead of plain number.
* Improve parsing logic that does not require double loop.

I'll open issues on the backend repo for possible future enhancements.

**Pull request merged**
